File: currency-service/src/main/java/br/edu/atitus/currencyservice/controllers/CurrencyController.java
package br.edu.atitus.currencyservice.controllers;

import br.edu.atitus.currencyservice.dtos.CurrencyDTO;
import br.edu.atitus.currencyservice.entities.CurrencyEntity;
import br.edu.atitus.currencyservice.infrastructure.exceptions.CurrencyNotFoundException;
import br.edu.atitus.currencyservice.repositories.CurrencyRepository;
import br.edu.atitus.currencyservice.services.CurrencyService;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/currency")
public class CurrencyController {

    private final CurrencyService currencyService;

    public CurrencyController(CurrencyService currencyService) {
        this.currencyService = currencyService;
    }

    @GetMapping(path = "/convert",params = {"source", "target"})
    public ResponseEntity<CurrencyDTO> findBySourceAndTarget(@RequestParam String source,
                                                            @RequestParam String target){

        CurrencyDTO dto = currencyService.findBySourceAndTarget(source,target);

        return ResponseEntity.ok().body(dto);
    }
}

File: currency-service-go/internal/controllers/currency_controller.go
package controllers

import (
	"encoding/json"
	"net/http"

	"github.com/go-chi/chi/v5"

	"github.com/PedroScheurer/currency-service/internal/apperrors"
	"github.com/PedroScheurer/currency-service/internal/services"
)

// CurrencyController é o equivalente Go do CurrencyController Java
// (@RestController @RequestMapping("/currency")).
type CurrencyController struct {
	service *services.CurrencyService
}

func NewCurrencyController(service *services.CurrencyService) *CurrencyController {
	return &CurrencyController{service: service}
}

func (c *CurrencyController) RegisterRoutes(r chi.Router) {
	r.Route("/currency", func(r chi.Router) {
		// Equivalente a @GetMapping(path = "/convert", params = {"source", "target"})
		r.Get("/convert", c.convert)
	})
}

// convert é o equivalente a CurrencyController.findBySourceAndTarget.
func (c *CurrencyController) convert(w http.ResponseWriter, r *http.Request) {
	source := r.URL.Query().Get("source")
	target := r.URL.Query().Get("target")

	if source == "" || target == "" {
		http.NotFound(w, r)
		return
	}

	dto, err := c.service.FindBySourceAndTarget(r.Context(), source, target)
	if err != nil {
		apperrors.WriteErrorResponse(w, err)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(http.StatusOK)
	_ = json.NewEncoder(w).Encode(dto)
}
